source/common/router/header_parser.cc

```diff
@@ -269,6 +270,18 @@ HeaderParser::configure(const Protobuf::RepeatedPtrField<HeaderValueOption>& hea
       append_action = header_value_option.append_action();
     }
 
+    // There should be only one OVERWRITE_IF_EXISTS_OR_ADD per each header.
```

Comment:
This change breaks backward compatibility, and rejecting it should be done with care (runtime guarding it, and allowing users and management server implementers update their code in a reasonable time).
To accept the config and to keep the semantics, can the same be achieved by linearizing the header_to_add list (i.e., ignoring the previous addition of the header key)?

Note that my suggestion assumes that a header cannot have circular reference to itself (IIRC before #21932 this was not allowed).

Reply:
Accepting the config with multiple `OVERWRITE_IF_EXISTS_OR_ADD` is easy (above and below lines from here should be deleted). It works fine, but CPU cycles will be wasted because whatever is calculated before the last occurrence of  `OVERWRITE_IF_EXISTS_OR_ADD` will be discarded and overwritten. So, to keep backwards compatibility for config I think that we should allow multiple `OVERWRITE_IF_EXISTS_OR_ADD` for the same headers.

As or semantics, I believe that what we are suggesting now is the cleanest (and probably compatible with what was before #21932, but I have to check). I think that when documentation uses words `IF_EXISTS` it means the headers in received packet, not "intermediate" headers created by for example `APPEND_IF_EXISTS_OR_ADD` when a new header is built.

Reply:
RE wasted CPU cycles, I agree. This can be addressed by linearizing the list of operations, and removing any operation that is overridden (iterating the list backwards, and maintaining the first encounter of each key).

I believe that the point you make about the semantics of `IF_EXISTS` is correct. I'm worried that rejecting a config that was valid before may cause some harm on current deployments that for some reason send a config that has multiple `OVERWRITE_IF_EXISTS_OR_ADD` actions for the same header.

Reply:
@adisuissa 

> RE wasted CPU cycles, I agree. This can be addressed by linearizing the list of operations, and removing any operation that is overridden (iterating the list backwards, and maintaining the first encounter of each key).

Let me do it as a separate PR. In case we need to revert it, it will be cleaner. Wasted CPU is a pre-existing issue, not related what we are addressing in this PR.

So, at the moment the code does not introduce any backward compatibility issues and if you do not see any other issues I think it it complete.